[fix] FIxed webworker-threads dependency

YanDoroshenko opened the pull request:
Webworker-threads dependency moved from devDependencies to dependencies because it is required for library to work

uruuru commented:
This issue is still on my agenda. I'm not really happy with the dependency since it is only required if one actually uses a web worker. However, since there's no notion of _optional_ dependencies in npm, I'm not sure if a better solution exists.

YanDoroshenko commented:
I've found something similar to optional dependencies. It requires a package to manage them though.

spoenemann commented:
> there's no notion of optional dependencies in npm

https://docs.npmjs.com/files/package.json#optionaldependencies

YanDoroshenko commented:
If I understand correctly, what @uruuru tried to say was not optional dependencies (which by NPM's definition are dependencies that may or may not be found), but rather on demand dependencies, which should be required only when necessary.

uruuru commented:
Regarding my interpretation of _optional_ Yan is correct. 

@YanDoroshenko thanks for your input. The _codependency_ package looks interesting. However, to avoid the dependency, I changed the code in a way that falls back to the non-worker version if the `worker-threads` package is unavailable and prints a warning.

YanDoroshenko commented:
Looks like it works fine, thanks.

YanDoroshenko commented:
Sorry, it does not work. I had a cached webworker-threads package. After cleanup it falls on being unable to find webworker-threads.

uruuru commented:
Did you use the version `0.0.2`? If so, how does it fail? For me it prints the desired warning if webworker-threads is missing.

YanDoroshenko commented:
I used `0.0.2`. Instead of warning it logs an error:
`Error: Cannot find module 'webworker-threads' from '.../src/main/webapp/node_modules/elkjs/lib'`

uruuru commented:
What's the exact line at which it fails?

YanDoroshenko commented:
No idea. How do I find out? NPM logger does not log stack trace or line number.
I think it's in `main.js`, line 23 or 30. You probably know that already though.

YanDoroshenko commented:
So I've added a `webworker-threads` dependency and added some logging to the `main.js` here (line 22):
```
try {
      require.resolve('webworker-threads');
      workerThreadsExist = true;
} catch (e) {
      console.log(e);
}
```
And I get a message that `require.resolve is not a function` in the console.

uruuru commented:
I should have asked this earlier I guess: what's your setup? Are you using node.js?

YanDoroshenko commented:
I am no JS guy so I don't know exactly what do you mean by "setup", but I have a webapp assembled with NPM running in the browser, . And I don't use node.js.

uruuru commented:
With setup I just mean the technologies you use. Since you're in the browser you don't need the webworkerthreads library anyway, and you should bypass the node.js related code in the `main-node.js`.

You should be able to do something like this:
```
var ELK = require('elkjs/lib/elk-api')
var elk = new ELK({
  workerUrl: './node_modules/elkjs/lib/elk-worker.min.js'
})
```

YanDoroshenko commented:
That does work perfectly. Thanks mate, you're some kind of a JavaScript wizard, I see.
